### PR TITLE
Speedup CI testing

### PR DESCRIPTION
**Pull request by oke-aditya**
## What does this PR do?

Speeds up CI Testing.

Some optmizations followed

- [x] Use `@torch.no_grad()` to avoid gradients.
- [x] Avoid Downloading Pretrained Weights, These are provided by torchvision we can rely on them and save time.
- [x] Use `fast_dev_run`, disable logging, checkpoint callbacks.

## Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? [not needed for typos/docs]
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning-bolts/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/models/self_supervised/test_resnets.py
Comment (oke-aditya) [inline tests/models/self_supervised/test_resnets.py:19]: Since for all the resnet models we don't compute backward. We shouldn't keep track of gradients, :smile:

File: tests/models/test_gans.py
Comment (oke-aditya) [inline tests/models/test_gans.py:22]: I think we can standardize to use something like this.

File: tests/models/test_vision.py
Comment (oke-aditya) [inline tests/models/test_vision.py:39]: No grad needed here too. (Might save a lot GPT-2 with no grad :thinking: )

File: tests/models/self_supervised/test_resnets.py
Comment (oke-aditya) [inline tests/models/self_supervised/test_resnets.py:39]: This can save a bunch of time, since it avoids totally a lot of models

**Comment by oke-aditya** [comment]
I think before merging, we can check if we are really getting speedup. 
I think it's very little around 2 mins or so in CI Full testing, latest. Disabling grads is giving slight benefit.

File: tests/models/test_gans.py
Comment (Borda) [reply tests/models/test_gans.py:22]: is it needed when you already have `fast_dev_run=True`?

File: tests/models/test_vision.py
Comment (Borda) [inline tests/models/test_vision.py:9]: this shall be inside functions

File: tests/models/test_vision.py
Comment (Borda) [inline tests/models/test_vision.py:60]:
```suggestion
def test_semantic_segmentation(tmpdir):
```

File: tests/models/test_vision.py
Comment (akihironitta) [inline tests/models/test_vision.py:11]: As long as we keep this seed, `test_igpt` passes with the current loss value `1.7`, right?

Comment (oke-aditya) [reply tests/models/test_vision.py:11]: Yes, but that's unreliable testing, I mean the loss should rather satisfy some value irrespective of seed.

File: tests/models/test_gans.py
Comment (oke-aditya) [reply tests/models/test_gans.py:22]: Well yes, This avoid some calls to logger, prevents checkpoints being stored.

Checkpoints can occupy unnecassary disk space and possible transfer from RAM to Memory causing `Memory Error`

I'm not sure if this will exactly solve the problem #526 But keeping this enabled will probably avoid these issues.

File: tests/models/test_vision.py
Comment (oke-aditya) [reply tests/models/test_vision.py:9]: Same comment as above. This is defensive programming, to avoid unnessary errors later :smiley:

Comment (oke-aditya) [reply tests/models/test_vision.py:9]: I tihnk globally seeding will avoid duplication of code and possible calls?
We can be determinstic with a seed say 0 or 42 (scikit-learn :smiley: number)

**Comment by oke-aditya** [comment]
I will change the issue name ready to merge, when optimization is good to go :smile: 

I will post a detailed comparison of optimization here, I will compare master runs with this.

File: tests/models/test_vision.py
Comment (akihironitta) [reply tests/models/test_vision.py:11]: Yeah, you're right... but I'm still not sure if specifying a certain value of the loss is reasonable because the value is based on the implementation which we want to test, right...? But for now, I guess we could just change this value.

**Comment by oke-aditya** [comment]
Hey @Borda  and @akihironitta , I have done optimizations and this is good to go.
I see no codecoverage dips, nor there is any CI failure :smile:  :rocket:

File: tests/models/test_gans.py
Comment (akihironitta) [reply tests/models/test_gans.py:22]: Do we need `max_epochs=1` here? `fast_dev_run=True` ends right after the first batch in the first epoch, right?

**Review by akihironitta — APPROVED** [review]
Other than @Borda's comment:
> is it needed when you already have `fast_dev_run=True`?

It looks good to me.

**Comment by oke-aditya** [comment]
Let me remove `max_epochs` with `fast_dev_run`

**Comment by oke-aditya** [comment]
@Borda  can you review :smiley:

File: tests/models/self_supervised/test_models.py
Comment (akihironitta) [inline tests/models/self_supervised/test_models.py:43]:
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/self_supervised/test_models.py
Comment (akihironitta) [inline tests/models/self_supervised/test_models.py:122]:
```suggestion
    trainer = pl.Trainer(gpus=0, fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_classic_ml.py
Comment (akihironitta) [inline tests/models/test_classic_ml.py:23]: Let's turn off saving checkpoints here since we're not using `fast_dev_run=True`. [the docs](https://pytorch-lightning.readthedocs.io/en/stable/debugging.html#fast-dev-run)

```suggestion
    trainer = pl.Trainer(max_epochs=300, default_root_dir=tmpdir, progress_bar_refresh_rate=0, logger=False, checkpoint_callback=False)
```

File: tests/models/test_classic_ml.py
Comment (akihironitta) [inline tests/models/test_classic_ml.py:54]: Same as above.
```suggestion
        logger=False,
        checkpoint_callback=False,
    )
```

File: tests/models/test_detection.py
Comment (akihironitta) [inline tests/models/test_detection.py:27]: I like your idea of being defensive, but lightning ensures to disable these options with `fast_dev_run=True` as stated in [the docs](https://pytorch-lightning.readthedocs.io/en/stable/debugging.html#fast-dev-run) so how about keeping this line as is? 
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_detection.py
Comment (akihironitta) [inline tests/models/test_detection.py:36]: Same as above.
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_gans.py
Comment (akihironitta) [inline tests/models/test_gans.py:21]:
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_gans.py
Comment (akihironitta) [inline tests/models/test_gans.py:37]:
```suggestion
    trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

File: tests/models/test_vision.py
Comment (akihironitta) [inline tests/models/test_vision.py:77]:
```suggestion
    trainer = trainer = pl.Trainer(fast_dev_run=True, default_root_dir=tmpdir)
```

**Review by akihironitta — COMMENTED** [review]
Hey @oke-aditya, I think this is a great improvement in the tests! The comments below are all from me, so once these are agreed upon between us, this PR is good to go :)

**Comment by oke-aditya** [comment]
Hey @akihironitta @Borda  I have done the changes :smiley: 

Sorry for the delay, last few weeks were tedious

**Review by akihironitta — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**
